src/coreclr/debug/ee/controller.cpp

```diff
@@ -3879,6 +3891,75 @@ void DebuggerController::DispatchMethodEnter(void * pIP, FramePointer fp)
 
 }
 
+void DebuggerController::EnableMultiCastDelegate()
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+    }
+    CONTRACTL_END;
+
+    ControllerLockHolder chController;
+    Debugger::DebuggerDataLockHolder chInfo(g_pDebugger);
+    if (!m_multicastDelegateHelper)
+    {
+        LOG((LF_CORDB, LL_INFO1000000, "DC::EnableMultiCastDel, this=%p, previously disabled\n", this));
+        m_multicastDelegateHelper = true;
+    }
+    else
+    {
+        LOG((LF_CORDB, LL_INFO1000000, "DC::EnableMultiCastDel, this=%p, already set\n", this));
+    }
+}
+
+void DebuggerController::DisableMultiCastDelegate()
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+    }
+    CONTRACTL_END;
+
+    ControllerLockHolder chController;
+    Debugger::DebuggerDataLockHolder chInfo(g_pDebugger);
+
+    if (m_multicastDelegateHelper)
+    {
+        LOG((LF_CORDB, LL_INFO10000, "DC::DisableMultiCastDel, this=%p, previously set\n", this));
+        m_multicastDelegateHelper = false;
+    }
+    else
+    {
+        LOG((LF_CORDB, LL_INFO10000, "DC::DisableMultiCastDel, this=%p, already disabled\n", this));
+    }
+}
+
+// Loop through controllers and dispatch TriggerMulticastDelegate
+void DebuggerController::DispatchMulticastDelegate(BYTE* pbDel, INT32 countDel)
+{
+    LOG((LF_CORDB, LL_INFO10000, "DC::DispatchMulticastDelegate\n"));
+
+    Thread * pThread = g_pEEInterface->GetThread();
+    _ASSERTE(pThread  != NULL);
+
+    ControllerLockHolder lockController;
+
+    DebuggerController *p = g_controllers;
+    while (p != NULL)
+    {
+        if (p->m_multicastDelegateHelper)
+        {
+            if ((p->GetThread() == NULL) || (p->GetThread() == pThread))
```

Comment:
Does this work correctly when there are multiple nested multicast delegate calls on the same thread?

Reply:
It needs some more work, thanks!

Reply:
```
Console.WriteLine("f"/*invocationList*/);
Action w = () => 
{ 
    Console.WriteLine("w"); 
};
Action x = () => 
{ 
    Console.WriteLine("x"); 
};
Action y = () => 
{
    Console.WriteLine("y");
};
Action z = w + x + y;
Action a = () => 
{ 
    Console.WriteLine("a"); 
    z();
};
Action b = () => 
{
    Console.WriteLine("b");
    z();
};
Action c = a + b;
c(); // step into me
Console.WriteLine("f"/*invocationList*/);
```
It works on this code. I can step into z in both a and b.